spawnpoint.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2021 Siôn le Roux.  All rights reserved.
+// Use of this source code is subject to an MIT-style
+// licence which can be found in the LICENSE file.
+
+package main
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/solarlune/resolv"
+)
+
+// spawnDistance is the distance where the point is activated
+const spawnDistance = 250
+
+// SpawnPoints is an array of SpawnPoint
+type SpawnPoints []*SpawnPoint
+
+// Update updates all the SpawnPoints
+func (sps *SpawnPoints) Update(g *Game) {
+	for _, s := range *sps {
+		s.Update(g)
+	}
+}
+
+// SpawnPosition describes the spawning position related to the SpawnPoint center
+type SpawnPosition struct {
+	Distance int // Distance of the position from the center
+	Angle    int // Angle of the position
+}
+
+// SpawnPoint is a point on the map where zombies are spawn
+type SpawnPoint struct {
+	Position       Coord
+	InitialCount   int
+	Continuous     bool
+	Zombies        Zombies
+	InitialSpawned bool
+	PrevPosition   SpawnPosition
+	NextSpawn      int
+}
+
+// NextPosition gives the offset of the next spawning to the center of the point
+func (s *SpawnPoint) NextPosition() Coord {
+	
+	// Move further if zombies have been spwaned around the whole circle
+	if s.PrevPosition.Angle == 0 {
+		s.PrevPosition.Distance++
+	}
+
+	// Spawn positions in every 60 degress (360 / 6)
+	s.PrevPosition.Angle = (s.PrevPosition.Angle + 1) % 6
+
+	return Coord{
+		X: math.Cos(-2*math.Pi / 6 * float64(s.PrevPosition.Angle)) * float64(s.PrevPosition.Distance),
+		Y: math.Sin(-2*math.Pi / 6 * float64(s.PrevPosition.Angle)) * float64(s.PrevPosition.Distance),
+	}
+}
+
+// SpawnZombie spawns one zombie
+func (s *SpawnPoint) SpawnZombie(g *Game) {
+	np := s.NextPosition()
+
+	zombieType := rand.Intn(zombieTypes)
+	speed := zombieSpeed
+	if zombieType == 2 {
+		speed = zombieRunnerSpeed
+	} else if zombieType == 3 {
+		speed = zombieCrawlerSpeed
+	}
+
+	sprites := g.ZombieSprites[zombieType]
+	dimensions := sprites.Sprite[0].Position
+	z := &Zombie{
+		Object: resolv.NewObject(
+			s.Position.X + np.X * 32, s.Position.Y + np.Y * 32,
```

Comment:
I haven't actually checked the maths, but is this big enough to guarantee that none of them will spawn on the same cell? Because then there's a chance of them getting stuck on each other.

Reply:
Yes, I tried with smaller offset, but they were stuck in the same cell then

Reply:
Alrighty then, should be good. I was looking at something like this previously: https://stackoverflow.com/questions/3706219/algorithm-for-iterating-over-an-outward-spiral-on-a-discrete-2d-grid-from-the-or but I don't have any preference to this or the other solution, they both look fine. The outwards spiral is obviously a lot a more complicated than this but it guarantees that they won't land on the same cell because it goes cell-by-cell, if that's already not an issue with your solution then it doesn't matter and feel free to ignore this.

One thing I just realised was also a problem with my old hard-coded solution: It doesn't check to see if the cell has a wall in it before spawning the zombie, it should skip those or it'll be stuck in it.

Reply:
Good point. I need to check the walls. I start to think that as we do not need too many zombies from one spawn point, hard coded solution could work nicely :)

Reply:
Check if spawning cell has a well: done